Feature/slash commands

JR-Mitchell opened the pull request:
**What?**
- Adds privileged commands allowing for reloading the guild tree and global tree
- Updates all games to be runnable only using slash commands, with fixed arguments
- Updates the dloader to be runnable either using message listening or slash commands, and provides suggestions when using slash commands

**Why?**
Embrace the new! This will mean you don't have to remember the e-bot's commands, at least for these ones.

JR-Mitchell committed:
feat: updates EBot to store the ID of an admin user, configurable via environment variables

JR-Mitchell committed:
feat: adds the dynamic_load tree branch, which reloads the tree if the user is the admin

JR-Mitchell committed:
feat: adds an additional dloader command for loading the global tree

JR-Mitchell committed:
feat: updates games to be called using slash commands

JR-Mitchell committed:
feat: updates the games to supply choices for the command argument

JR-Mitchell committed:
feat: updates the dynamic loader to be a hybrid command with autocomplete suggestions

fjebaker commented on src/cogs/dynamic_load.py:
```diff
@@ -57,23 +62,45 @@ def _fmt_cog_list(self, input_list: list) -> str:
             await context.send(f"Cogs currently loaded:\n{resp}")
 
+        elif cog_name == "tree":
+            if (str(context.message.author.id) == self.bot.admin_user):
```
This looks like there can only be one admin user which isn't very democratic :(

fjebaker commented on src/__main__.py:
```diff
@@ -9,5 +9,5 @@
     logging.info("Starting ebot...")
 
-    bot = EBot()
+    bot = EBot(os.environ["ADMIN_USER"])
```
List of users?

fjebaker commented on src/cogs/dynamic_load.py:
```diff
@@ -57,23 +62,45 @@ def _fmt_cog_list(self, input_list: list) -> str:
 
+        elif cog_name == "tree":
+            if (str(context.message.author.id) == self.bot.admin_user):
+                self.bot.tree.copy_global_to(guild = context.guild)
+                await self.bot.tree.sync(guild = context.guild)
```
I don't understand what these things are doing. I get that `gtree` is likely `guild tree` but an inline comment explaining this would be useful.

JR-Mitchell replied:
Part of the issue is that I also don't fully understand what they are doing 😅

JR-Mitchell replied:
@fjebaker 
For more clarity,
I had a hard time getting the slash commands to work, even when following online advice and tutorials. In the end I used both what `tree` and `gtree` do, not sure in what order, so I don't understand fully.
However, as of 23c4697f57fbfd6d90e0d33023a5c187de1149ea I've renamed them to `globaltree` and `guildtree` so that the difference is clearer.
My (probably wrong) understanding of it is:
- `globaltree` updates the global configuration for the command tree for the bot, including what slash commands can be called, what arguments those slash commands can have, any input validation / autosuggest for those arguments, etc. But, it can take a while for the global configuration update to be reflected on all servers
- `guildtree` copies the global configuration to the configuration of a single guild (i.e the one that the command is called on), and then synchronises for that guild, which is a much faster configuration change and so the update will be on that server faster
There's also apparently heavy rate limiting, so these should only be called if there is definitely a change in the command tree structures that needs to be updated, hence why these are privileged commands.

fjebaker commented on src/ebot.py:
```diff
@@ -20,6 +20,7 @@ def __init__(self):
             intents=intents,
         )
         self.logging = logging.getLogger(__name__)
+        self.admin_user = admin_user
```
`self.admin_users` comrade... :(

fjebaker reviewed (commented):
LGTM but maybe multiple admin users?

fjebaker commented:
Please run linter etc. as well

JR-Mitchell committed:
feat(admin): updates ebot so that a list of admin ids is passed from the environment

JR-Mitchell committed:
chore: fixes some lint issues

JR-Mitchell commented:
> LGTM but maybe multiple admin users?

a781a60e80e3aa36dc48808656048db3e4bf0969 resolves this

JR-Mitchell committed:
feat: further linting fixes

JR-Mitchell committed:
chore: upgrade linter.yml

Adds import for emoji

JR-Mitchell committed:
chore: even more linting fixes

JR-Mitchell commented:
> Please run linter etc. as well

I've fixed a bunch of linter issues, but many were from previous code. I'm pretty much tapped on trying to resolve the last few.
Also, python style baffles me.
Why is:
```
if (context.message.author.id in self.bot.admin_users):
```
considered worse than
```
if context.message.author.id in self.bot.admin_users:
```
?
Imo, the first is much more readable.

JR-Mitchell committed:
chore: update test-pr.yml

Adds the emoji dependency so that the test doesn't fail

fjebaker committed:
fix: remove gpt and linter fixes

Addressed all pylint and flake issues, and removed the GPT cog as this
is no longer something we are maintaining.

fjebaker committed:
format: ran black over the src directory

fjebaker commented on .github/workflows/test-pr.yml:
```diff
@@ -21,7 +21,7 @@ jobs:
       run: |
         python -m pip install --upgrade pip
-        pip install pytest pytest-cov pytest-asyncio discord chatgpt
+        pip install pytest pytest-cov pytest-asyncio discord chatgpt emoji
```
Same here, remove chatgpt 
```suggestion
        pip install pytest pytest-cov pytest-asyncio discord emoji
```

JR-Mitchell replied:
I think doing that would break the tests again. Imo, it's a separate PR to remove the gpt cog completely

fjebaker replied:
Fair enough

fjebaker commented on .github/workflows/linter.yml:
```diff
@@ -24,7 +24,7 @@ jobs:
       run: |
         python -m pip install --upgrade pip
-        pip install flake8 pylint discord chatgpt
+        pip install flake8 pylint discord chatgpt emoji
```
```suggestion
        pip install flake8 pylint discord emoji
```

fjebaker commented on src/__main__.py:
```diff
@@ -9,5 +10,7 @@
     logging.info("Starting ebot...")
 
-    bot = EBot()
+    admins = json.loads(os.environ["ADMIN_USERS"])
```
Uhhh does that mean you have to do something like
```
ADMIN_USERS='["name1", "name2"]' python start-ebot
```
??

JR-Mitchell replied:
With `pipenv` you can have a `.env` file in the root directory (which is ignored by the `.gitignore`), and so when you `pipenv shell` it'll load the env vars from that file. So it'd look like:
```
DISCORD_TOKEN=tokenGoesHere
ADMIN_USERS=[12345,67891]
```

fjebaker replied:
Ah nice okay

fjebaker commented on src/ebot.py:
```diff
@@ -10,7 +10,7 @@
     """TODO"""
 
-    def __init__(self):
+    def __init__(self, admin_users):
```
Type annotation missing

JR-Mitchell replied:
Done on 695da1f725808b745fd14c14cf227aeaf2c3e6d6

fjebaker commented on src/games/ridethebus.py:
```diff
@@ -102,8 +102,7 @@ def sparse(self) -> str:
         self.current_row = 1
-        for c in itertools.chain(*self._cards):
-            yield c
+        yield from itertools.chain(*self._cards)
```
TIL new syntax

JR-Mitchell committed:
chore: adds type annotation to EBot constructor

JR-Mitchell committed:
feat: makes the dloader tree commands more clearly named

fjebaker reviewed (approved):
:shipit:
